django/db/migrations/autodetector.py

```diff
@@ -52,21 +52,33 @@ def deep_deconstruct(self, obj):
         Used for full comparison for rename/alter; sometimes a single-level
         deconstruction will not compare correctly.
         """
-        if not hasattr(obj, 'deconstruct') or isinstance(obj, type):
-            return obj
-        deconstructed = obj.deconstruct()
-        if isinstance(obj, models.Field):
-            # we have a field which also returns a name
-            deconstructed = deconstructed[1:]
-        path, args, kwargs = deconstructed
-        return (
-            path,
-            [self.deep_deconstruct(value) for value in args],
-            {
+        if isinstance(obj, list):
+            items = [self.deep_deconstruct(value) for value in obj]
+        elif isinstance(obj, tuple):
+            items = tuple(self.deep_deconstruct(value) for value in obj)
+        elif isinstance(obj, dict):
+            items = {
                 key: self.deep_deconstruct(value)
-                for key, value in kwargs.items()
-            },
-        )
+                for key, value in obj.items()
+            }
+        elif isinstance(obj, type):
```

Comment:
https://code.djangoproject.com/ticket/22951 is the relevant issue for this - I can confirm that the test from that issue (`migrations.test_autodetector.AutodetectorTests.test_deconstruct_type`) blows up when this branch is omitted. Comment added in 4204a561bdc6bfa185665256b956954b3499530d